Core: Fix builder stats typings to be optional

AriPerkkio opened the pull request:
Issue: https://github.com/storybookjs/builder-vite/issues/397

## What I did

Fixed typings of `Builder` (`@storybook/core-common` package). Currently typings and implementation do not match. The implementation is using `@ts-ignore` 😭

https://github.com/storybookjs/storybook/blob/61d6451c4bfc7eb86f1f7ea2ee6d64b9a5d3a976/lib/core-server/src/build-dev.ts#L109-L111

Usage of `Builder<SomeValidConfigType, {}>` causes TypeErrors mentioned in the linked issue. If typings were correct these kind of issues would be prevented.  

## How to test

- [ ] Is this testable with Jest or Chromatic screenshots?
- [ ] Does this need a new example in the kitchen sink apps?
- [ ] Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `master` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/react/contribute/how-to-contribute

Maintainers: Please tag your pull request with at least one of the following:
`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

-->

AriPerkkio committed:
fix: builder stats typings to be optional

ndelangen commented on lib/core-common/src/types.ts:
```diff
@@ -234,22 +234,22 @@ export interface StorybookConfigOptions {
 export type Options = LoadOptions & StorybookConfigOptions & CLIOptions & BuilderOptions;
 
-export interface Builder<Config, Stats> {
+export interface Builder<Config, BuilderStats extends Stats | null> {
```
can you explain the `| null` part? everything else makes sense to me

AriPerkkio replied:
The `null` part here is so that custom builders can disable use of `stats`. Here is how `builder-vite` is utilizing this interface: https://github.com/storybookjs/builder-vite/blob/6338427ebaf4152ae098f5a06dc065bf774d9186/packages/builder-vite/index.ts#L14-L17.

As this is a required parameter of interface generics, `null` is used to disable it. Should we make it optional, and default to `Stats`? Also make `start` method's output `{ stats }` optional:

```diff
-export interface Builder<Config, BuilderStats extends Stats | null> {
+export interface Builder<Config, BuilderStats extends Stats = Stats> {
    getConfig: (options: Options) => Promise<Config>;
    start: (args: {
        options: Options;
        startTime: ReturnType<typeof process.hrtime>;
        router: Router;
        server: Server;
    }) => Promise<void | {
-       stats: BuilderStats;
+       stats?: BuilderStats;
        totalTime: ReturnType<typeof process.hrtime>;
        bail: (e?: Error) => Promise<void>;
    }>;
```

Now custom builders could simply ignore the stats when using `Builder` interface:

```ts
export type ViteBuilder = Builder<UserConfig>;
```

AriPerkkio commented:
> It also seems that this causes 2 issues:
> [...] `Type 'Stats' is missing the following properties from type 'Stats': formatFilePath, normalizeFieldKey, sortOrderRegular`

Good catch. I was unable to run storybook locally when creating the PR but got it running now - I'm also seeing this error now.

Looks like it's caused by Webpack v4 vs v5 `Stats` typings. Is there a way to have two versions of webpack's typings in `@storybook/core-common`? Then simply use `type BuilderStats = Webpack4Stats | Webpack5Stats;`. Currently the types are coming from here:

https://github.com/storybookjs/storybook/blob/008b54258be41f87ba45abd25882b83f131d0ec1/lib/core-common/package.json#L92

https://github.com/storybookjs/storybook/blob/008b54258be41f87ba45abd25882b83f131d0ec1/lib/core-common/src/types.ts#L4

Another option is to create a custom interface where we explicitly list the types we expect from a builder stats. This would mean that whole codebase should be scanned for the usage of stats - as there are `@ts-ignore`'s used, these won't come up automatically, I think.
Something like:

```ts
/** Webpack 4/5 compatible `Stats` interface */
interface Stats {
    /** Returns compilation information as a JSON object. */
    toJson(options?: SomeType, forToString?: boolean): SomeOutputType;
}
```

ndelangen commented:
@AriPerkkio I've been working on `future/base` and a series of PRs that will all be part of `7.0`..
The whole webpack `4` & `5` support and having it in 1 monorepo is a pain, believe me I know... this (part of)is why in `7.0` we'll drop webpack `4` support.

Perhaps you'll be keen to take a look at my work here:
https://github.com/storybookjs/storybook/pull/18504

Do you think it'd be best to focus your efforts on `7.0` as well?

FYI: The default branch `next` is in a in between state between `6.5` and `7.0`, where most work will become part of `7.0` anyway.
But it helps us keep merge conflicts to a minimum so we *can* cherry-pick important fixes to `main` and thus release in `6.5.x`

AriPerkkio commented:
> Do you think it'd be best to focus your efforts on `7.0` as well?

Yes, I think we should only target the `7.0` here and Webpack 5 only. It makes thing a lot easier.

However it seems that in https://github.com/storybookjs/storybook/pull/18504 the `@storybook/core-common` does not have Webpack typings anymore. There is a new package, `@storybook/core-webpack`, which handles those. `@storybook/core-common` is still the one that defines `Builder`
s `stats`: https://github.com/storybookjs/storybook/blob/9b92429b797a777fbba837b3b310529a22580911/lib/core-common/src/types.ts#L193. 
The idea of separating Webpack typings from `@storybook/core-common` is great but those `stats` are still a Webpack thing. I don't think we want to make `@storybook/core-common` dependent of `@storybook/webpack-common`, right? Where should the typings of `stats` we defined?

ndelangen commented:
They should be super generic in core-common then be extended and made webpack specific in core-webpack, and then possibly be extended again in builder-webpack5.

ndelangen commented:
@AriPerkkio let's rebase this unto `future/base` and proceed from there? Would you need my help with that?

AriPerkkio commented:
Rebased `future/base` and changed that one to be PR target branch. It already seems to have generic typings for `Stats` - this is perfect:

https://github.com/storybookjs/storybook/blob/379d8b83bdb88ded019fd2720ef7ad94ca0dfc55/lib/core-common/src/types.ts#L102-L104
 
Now, with changes of this PR, builders which do not provide such `Stats`, can utilize the `Builder` interface without providing dummy `Stats`, e.g. `Builder<ViteConfig>`. I also removed the `@ts-ignore`'s from the code now that everything is properly typed. There should be no more null pointers caused by falsely typed builders!

ndelangen commented:
Awesome, I'll take a look on Monday!

ndelangen committed:
Merge branch 'future/base' into fix/builder-stats-typings

ndelangen commented:
Sorry for the delay, I merged in the base branch, to see if the CI approves... If it does, I'm going to merge it! 👏👏👏
